File: src/main/java/org/taktik/icure/entities/Role.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.entities;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.taktik.icure.constants.Roles;
import org.taktik.icure.entities.base.Principal;
import org.taktik.icure.entities.base.StoredDocument;
import org.taktik.icure.entities.embed.Permission;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Role extends StoredDocument implements Principal, Cloneable, Serializable {
	private static final long serialVersionUID = 1L;

	protected String name;

	protected Roles.VirtualHostDependency virtualHostDependency;

	protected Set<Property> properties = new HashSet<>();

	protected Set<Permission> permissions = new HashSet<>();

	protected Set<String> children = new HashSet<>();
	protected Set<String> parents = new HashSet<>();
	protected Set<String> users = new HashSet<>();
	protected Set<String> virtualHosts = new HashSet<>();

	public Set<String> getChildren() {
		return children;
	}

	public void setChildren(Set<String> value) {
		this.children = value;
	}

	public Set<String> getParents() {
		return parents;
	}

	public void setParents(Set<String> value) {
		this.parents = value;
	}

	public Set<String> getUsers() {
		return users;
	}

	public void setUsers(Set<String> users) {
		this.users = users;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	@Override
	public Roles.VirtualHostDependency getVirtualHostDependency() {
		return virtualHostDependency;
	}

	public void setVirtualHostDependency(Roles.VirtualHostDependency virtualHostDependency) {
		this.virtualHostDependency = virtualHostDependency;
	}

	public Set<Permission> getPermissions() {
		return permissions;
	}

	public void setPermissions(Set<Permission> permissions) {
		this.permissions = permissions;
	}

	public Set<Property> getProperties() {
		return properties;
	}

	public void setProperties(Set<Property> properties) {
		this.properties = properties;
	}

	@Override
	public Set<String> getVirtualHosts() {
		return virtualHosts;
	}

	public void setVirtualHosts(Set<String> virtualHosts) {
		this.virtualHosts = virtualHosts;
	}

	@Override
	public Object clone() throws CloneNotSupportedException {
		return super.clone();
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		Role role = (Role) o;

		if (id != null ? !id.equals(role.id) : role.id != null) return false;

		return true;
	}

	@Override
	public int hashCode() {
		return id != null ? id.hashCode() : 0;
	}
}

File: src/main/kotlin/org/taktik/icure/entities/Role.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.entities

import java.io.Serializable
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.pozo.KotlinBuilder
import org.taktik.couchdb.entity.Attachment
import org.taktik.icure.entities.base.PropertyStub
import org.taktik.icure.entities.base.StoredDocument
import org.taktik.icure.entities.embed.RevisionInfo
import org.taktik.icure.entities.security.Permission
import org.taktik.icure.entities.security.Principal
import org.taktik.icure.utils.DynamicInitializer
import org.taktik.icure.utils.invoke

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class Role(
	@JsonProperty("_id") override val id: String,
	@JsonProperty("_rev") override val rev: String? = null,
	@JsonProperty("deleted") override val deletionDate: Long? = null,

	override val name: String? = null,
	override val properties: Set<PropertyStub> = emptySet(),
	override val permissions: Set<Permission> = emptySet(),
	private val parents: Set<String> = emptySet(),

	@JsonProperty("_attachments") override val attachments: Map<String, Attachment>? = emptyMap(),
	@JsonProperty("_revs_info") override val revisionsInfo: List<RevisionInfo>? = emptyList(),
	@JsonProperty("_conflicts") override val conflicts: List<String>? = emptyList(),
	@JsonProperty("rev_history") override val revHistory: Map<String, String>? = emptyMap()

) : StoredDocument, Principal, Cloneable, Serializable {
	companion object : DynamicInitializer<Role>

	fun merge(other: Role) = Role(args = this.solveConflictsWith(other))
	fun solveConflictsWith(other: Role) = super.solveConflictsWith(other) + mapOf(
		"name" to (this.name ?: other.name),
		"properties" to (other.properties + this.properties),
		"permissions" to (other.permissions + this.permissions),
		"parents" to (other.parents + this.parents)
	)

	override fun withIdRev(id: String?, rev: String) = if (id != null) this.copy(id = id, rev = rev) else this.copy(rev = rev)
	override fun withDeletionDate(deletionDate: Long?) = this.copy(deletionDate = deletionDate)
	override fun getParents(): Set<String> = parents
}
